### DISCUSSION TITLE
"throws" isn't catching the error

### DISCUSSION THREAD

**Discussion by jmls** [avajs/ava#3070] [General]
I know that I must be doing something wrong but for the life of me I can;t see it. I want to test this method of a class:

```
download(options: { id: string }) {
        const { id } = options;
        console.log('#A');
        console.log('#B');

        try {
            console.log('#C');
            // return true;
            const file = createReadStream(`/workspace/data/backup/${id}.sql`);

            return new StreamableFile(file);
        } catch (e) {
            throw new Error(`invalid backup id ${id}`);
        }
    }
```
  
  I'm expecting the     `createReadStream` to throw an error because the id is invalid. 
  
  My test is this
  
  ```
test.only('download should fail with invalid id', async (t: Assertions) => {
    const id = uuid();

    t.throws(
        () => {
            console.log('####1');
            service.download({
                id
            });
            console.log('####2');
            return true;
        },
        {
            message: (message) => ['🦄', '🐴'].includes(message)
        }
    );
});
```

now for the strange part. this is the console output

```
####1
  ✖ download should fail with invalid id 

  Uncaught exception in src/server/backup/_test/backup.spec.ts

  Error: ENOENT: no such file or directory, open '/workspace/data/backup/035bed6a-f0f1-4d96-9b2e-bbcc618fa925.sql'

[Error: ENOENT: no such file or directory, open '/workspace/data/backup/035bed6a-f0f1-4d96-9b2e-bbcc618fa925.sql'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/workspace/data/backup/035bed6a-f0f1-4d96-9b2e-bbcc618fa925.sql'
} Uncaught Exception thrown
  ✖ src/server/backup/_test/backup.spec.ts exited with a non-zero exit code: 1
  ─

  download should fail with invalid id

  src/server/backup/_test/backup.spec.ts:254

   253:                
   254:     t.throws(  
   255:         () => {

  Function returned:

  true
  
```
  notice that there are no #A or #B or #C in the output. It also looks like the error in the catch hasn't been thrown either
  
  if I uncomment the "return true" in the "try" this is the log
  
```
  ####1
  ✖ download should fail with invalid id 
#A
#B
#C
####2
  ─

  download should fail with invalid id

  src/server/backup/_test/backup.spec.ts:254

   253:                
   254:     t.throws(  
   255:         () => {

  Function returned:

  true

  › src/server/backup/_test/backup.spec.ts:254:7

  ─

  1 test failed
error Command failed with exit code 1.

```
this time I do get the #A, #B and #C 

what is going on ? Have I completely lost the plot ?

**Comment by novemberborn** [comment]
The error is emitted asynchronously as an event, definitely on the file stream and perhaps by that wrapper class. It's not synchronous so this code won't catch it. Neither will the `throwsAsync` assertion since that only works with promises.

Look into error handling with Node.js streams for more context.

**Reply by jmls** [reply]
Omg it's so bloody obvious 😂 thank you, sorry for the waste of bandwidth.

**Comment by jmls** [comment]
nope. Still struggling with this. Guess I need an idiot's guide  

Given this statement : `const file = createReadStream(`/workspace/data/backup/${id}.sql`);` how would I 

- catch the error when the file does not exist
- convert the error into a "nice" error message
- test in ava that the appropriate error has been thrown

**Reply by novemberborn** [reply]
The stream will emit an error event. You could perhaps do something like this:

```js
const {once} = require('events');
const file = createReadStream(`/workspace/data/backup/${id}.sql`);
await t.throwsAsync(once(file, 'end'));
```
